File: dawn/src/main/java/com/lzq/dawn/util/sdcard/SDCardUtils.java
package com.lzq.dawn.util.sdcard;

import android.content.Context;
import android.os.Environment;
import android.os.storage.StorageManager;
import android.os.storage.StorageVolume;

import com.lzq.dawn.DawnBridge;

import java.lang.reflect.Array;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;

/**
 * @Name :SDCardUtils
 * @Time :2022/8/29 15:49
 * @Author :  Lzq
 * @Desc : SD卡
 */
public final class SDCardUtils {
    private SDCardUtils() {
    }

    /**
     * 返回SD卡是否启用。
     *
     * @return {@code true}: enabled<br>{@code false}: disabled
     */
    public static boolean isSDCardEnableByEnvironment() {
        return Environment.MEDIA_MOUNTED.equals(Environment.getExternalStorageState());
    }

    /**
     * 返回SD卡的路径。
     *
     * @return SD卡的路径。
     */
    public static String getSDCardPathByEnvironment() {
        if (isSDCardEnableByEnvironment()) {
            return Environment.getExternalStorageDirectory().getAbsolutePath();
        }
        return "";
    }

    /**
     * 返回SD卡的信息。
     *
     * @return SD卡的信息。
     */
    public static List<SDCardInfo> getSDCardInfo() {
        List<SDCardInfo> paths = new ArrayList<>();
        StorageManager sm = (StorageManager) DawnBridge.getApp().getSystemService(Context.STORAGE_SERVICE);
        if (sm == null) {
            return paths;
        }
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N) {
            List<StorageVolume> storageVolumes = sm.getStorageVolumes();
            try {
                //noinspection JavaReflectionMemberAccess
                Method getPathMethod = StorageVolume.class.getMethod("getPath");
                for (StorageVolume storageVolume : storageVolumes) {
                    boolean isRemovable = storageVolume.isRemovable();
                    String state = storageVolume.getState();
                    String path = (String) getPathMethod.invoke(storageVolume);
                    paths.add(new SDCardInfo(path, state, isRemovable));
                }
            } catch (NoSuchMethodException e) {
                e.printStackTrace();
            } catch (IllegalAccessException e) {
                e.printStackTrace();
            } catch (InvocationTargetException e) {
                e.printStackTrace();
            }
        } else {
            try {
                Class<?> storageVolumeClazz = Class.forName("android.os.storage.StorageVolume");
                //noinspection JavaReflectionMemberAccess
                Method getPathMethod = storageVolumeClazz.getMethod("getPath");
                Method isRemovableMethod = storageVolumeClazz.getMethod("isRemovable");
                //noinspection JavaReflectionMemberAccess
                Method getVolumeStateMethod = StorageManager.class.getMethod("getVolumeState", String.class);
                //noinspection JavaReflectionMemberAccess
                Method getVolumeListMethod = StorageManager.class.getMethod("getVolumeList");
                Object result = getVolumeListMethod.invoke(sm);
                final int length = Array.getLength(result);
                for (int i = 0; i < length; i++) {
                    Object storageVolumeElement = Array.get(result, i);
                    String path = (String) getPathMethod.invoke(storageVolumeElement);
                    boolean isRemovable = (Boolean) isRemovableMethod.invoke(storageVolumeElement);
                    String state = (String) getVolumeStateMethod.invoke(sm, path);
                    paths.add(new SDCardInfo(path, state, isRemovable));
                }
            } catch (ClassNotFoundException e) {
                e.printStackTrace();
            } catch (InvocationTargetException e) {
                e.printStackTrace();
            } catch (NoSuchMethodException e) {
                e.printStackTrace();
            } catch (IllegalAccessException e) {
                e.printStackTrace();
            }
        }
        return paths;
    }

    /**
     * 返回已安装SD卡的path。
     *
     * @return 已安装SD卡的path。
     */
    public static List<String> getMountedSDCardPath() {
        List<String> path = new ArrayList<>();
        List<SDCardInfo> sdCardInfo = getSDCardInfo();
        if (sdCardInfo == null || sdCardInfo.isEmpty()) {
            return path;
        }
        for (SDCardInfo cardInfo : sdCardInfo) {
            String state = cardInfo.getState();
            if (state == null) {
                continue;
            }
            if ("mounted".equals(state.toLowerCase())) {
                path.add(cardInfo.getPath());
            }
        }
        return path;
    }


    /**
     * 返回外部存储的总大小
     *
     * @return 外部存储的总大小
     */
    public static long getExternalTotalSize() {
        return DawnBridge.getFsTotalSize(getSDCardPathByEnvironment());
    }

    /**
     * Return the available size of external storage.
     *
     * @return the available size of external storage
     */
    public static long getExternalAvailableSize() {
        return DawnBridge.getFsAvailableSize(getSDCardPathByEnvironment());
    }

    /**
     * Return the total size of internal storage
     *
     * @return the total size of internal storage
     */
    public static long getInternalTotalSize() {
        return DawnBridge.getFsTotalSize(Environment.getDataDirectory().getAbsolutePath());
    }

    /**
     * Return the available size of internal storage.
     *
     * @return the available size of internal storage
     */
    public static long getInternalAvailableSize() {
        return DawnBridge.getFsAvailableSize(Environment.getDataDirectory().getAbsolutePath());
    }


}

File: dawn/src/main/java/com/lzq/dawn/util/sdcard/SDCardUtils.kt
package com.lzq.dawn.util.sdcard

import android.content.Context
import android.os.Build
import android.os.Environment
import android.os.storage.StorageManager
import android.os.storage.StorageVolume
import com.lzq.dawn.DawnBridge
import java.lang.reflect.Array
import java.lang.reflect.InvocationTargetException
import java.util.Locale

/**
 * @Name :SDCardUtils
 * @Time :2022/8/29 15:49
 * @Author :  Lzq
 * @Desc : SD卡
 */
object SDCardUtils {
    @JvmStatic
    val isSDCardEnableByEnvironment: Boolean
        /**
         * 返回SD卡是否启用。
         *
         * @return `true`: enabled<br></br>`false`: disabled
         */
        get() = Environment.MEDIA_MOUNTED == Environment.getExternalStorageState()
    @JvmStatic
    val sDCardPathByEnvironment: String
        /**
         * 返回SD卡的路径。
         *
         * @return SD卡的路径。
         */
        get() = if (isSDCardEnableByEnvironment) {
            Environment.getExternalStorageDirectory().absolutePath
        } else ""

    @JvmStatic
    val sDCardInfo: List<SDCardInfo>
        /**
         * 返回SD卡的信息。
         *
         * @return SD卡的信息。
         */
        get() {
            val paths: MutableList<SDCardInfo> = ArrayList()
            val sm = DawnBridge.getApp().getSystemService(Context.STORAGE_SERVICE) as StorageManager
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                val storageVolumes = sm.storageVolumes
                try {
                    val getPathMethod = StorageVolume::class.java.getMethod("getPath")
                    for (storageVolume in storageVolumes) {
                        val isRemovable = storageVolume.isRemovable
                        val state = storageVolume.state
                        val path = getPathMethod.invoke(storageVolume) as String
                        paths.add(SDCardInfo(path, state, isRemovable))
                    }
                } catch (e: NoSuchMethodException) {
                    e.printStackTrace()
                } catch (e: IllegalAccessException) {
                    e.printStackTrace()
                } catch (e: InvocationTargetException) {
                    e.printStackTrace()
                }
            } else {
                try {
                    val storageVolumeClazz = Class.forName("android.os.storage.StorageVolume")
                    val getPathMethod = storageVolumeClazz.getMethod("getPath")
                    val isRemovableMethod = storageVolumeClazz.getMethod("isRemovable")
                    val getVolumeStateMethod = StorageManager::class.java.getMethod("getVolumeState", String::class.java)
                    val getVolumeListMethod = StorageManager::class.java.getMethod("getVolumeList")
                    val result = getVolumeListMethod.invoke(sm)
                    val length = Array.getLength(result)
                    for (i in 0 until length) {
                        val storageVolumeElement = Array.get(result, i)
                        val path = getPathMethod.invoke(storageVolumeElement) as String
                        val isRemovable = isRemovableMethod.invoke(storageVolumeElement) as Boolean
                        val state = getVolumeStateMethod.invoke(sm, path) as String
                        paths.add(SDCardInfo(path, state, isRemovable))
                    }
                } catch (e: ClassNotFoundException) {
                    e.printStackTrace()
                } catch (e: InvocationTargetException) {
                    e.printStackTrace()
                } catch (e: NoSuchMethodException) {
                    e.printStackTrace()
                } catch (e: IllegalAccessException) {
                    e.printStackTrace()
                }
            }
            return paths
        }
    @JvmStatic
    val mountedSDCardPath: List<String>
        /**
         * 返回已安装SD卡的path。
         *
         * @return 已安装SD卡的path。
         */
        get() {
            val path: MutableList<String> = ArrayList()
            val sdCardInfo = sDCardInfo
            if (sdCardInfo.isEmpty()) {
                return path
            }
            for (cardInfo in sdCardInfo) {
                val state = cardInfo.state ?: continue
                if ("mounted" == state.lowercase(Locale.getDefault())) {
                    path.add(cardInfo.path)
                }
            }
            return path
        }

    @JvmStatic
    val externalTotalSize: Long
        /**
         * 返回外部存储的总大小
         *
         * @return 外部存储的总大小
         */
        get() = DawnBridge.getFsTotalSize(sDCardPathByEnvironment)

    @JvmStatic
    val externalAvailableSize: Long
        /**
         * Return the available size of external storage.
         *
         * @return the available size of external storage
         */
        get() = DawnBridge.getFsAvailableSize(sDCardPathByEnvironment)

    @JvmStatic
    val internalTotalSize: Long
        /**
         * Return the total size of internal storage
         *
         * @return the total size of internal storage
         */
        get() = DawnBridge.getFsTotalSize(Environment.getDataDirectory().absolutePath)

    @JvmStatic
    val internalAvailableSize: Long
        /**
         * Return the available size of internal storage.
         *
         * @return the available size of internal storage
         */
        get() = DawnBridge.getFsAvailableSize(Environment.getDataDirectory().absolutePath)
}